README.md
# Node URL Permission

This Node.js library facilitates formatting permissions for users, groups or any other security principal in the following way:

```
<path>?<parameters>:<privileges>
```

**For example:**

```js
import { permissions } from 'url-permissions';

function authorize(userPermissions) {
  if (!permissions(userPermissions).allows('/articles:read')) {
    throw new Error('Not allowed here!')
  }
}

// Permission '/articles:read,create' will grant access to '/articles:read'
authorize(['/articles:read', '/accounts/12345:owner', '/articles?author=12345:owner']);
```

# Description

URL Permissions are intended to provide authorization for web services that:

1. have a REST API;
2. require permissions to be expressed in a succinct way;
3. have complex authorization requirements.

Read more about the [URL Permission format](https://github.com/nielskrijger/url-permissions).

## Format

```
<path>?<parameters>:<privileges>
```

An URL Permission consists of three components:

1. **Path**. The path identifies which resource the permission applies to. Paths can be absolute pathnames (starting with `/`) or an entire url with domain name and url scheme.

    Paths may include wildcards `_` (single character) and `*` (any or zero characters).

    For example, to read a user comment accessible at `https://api.example.com/articles/article-1/comments/comment-1` you might use one of the following URL permissions:

    ```
    /articles/*:read
    /articles/*/comments/*:read
    https://api.example.com/articles/*:read
    https://api.example.com/articles/article-1/comments/comment-1:read
    ```

2. **Parameters**. Parameters are optional attribute-based restrictions and are  very similar to url query parameters.

    For example, the permission `/articles:read` grants read access to all articles whereas:

    ```
    /articles?author=user-1:read
    ```

    ... grants read access only to articles whose author is `user-1`.

    You can specify multiple parameters and values like so:

    ```
    /articles?author=user-1,user-2&status=published:read
    ```

    ... which allows read access of published articles whose author is `user-1` or `user-2` (comma-separated values are evaluated as OR).

3. **Privileges**. Privileges specify which operations are allowed on the resource. You can either specify these as a comma-separated set of names, a bitmask, or a combination thereof (e.g. `create,read,update,delete`, `15` and `crud` are equivalent). Privileges are fully customizable. The default are:

    Bitmask | Name            | Description
    --------|-----------------|-----------------
    `1`     | `read`          | View resource.
    `2`     | `create`        | Create resource.
    `4`     | `update`        | Update resource.
    `8`     | `delete`        | Delete resource.
    `15`    | `crud`          | All of the above.
    `16`    | `manage`        | Set permissions of subject without `manage` or `super` privilege.
    `31`    | `manager`       | All of the above.
    `32`    | `own`           | Set permissions of subjects without `super` privilege + allow ownership transfer.
    `63`    | `owner`         | All of the above.
    `64`    | `admin`         | Enables special administrative actions and setting permissions of everyone.
    `127`   | `administrator` | All of the above.

# permission(perm)

The `permission(perm)` function enables a variety of evaluation and transformation methods. To evaluate a collection of permissions scroll down until you hit the `permissions(perms[])` section.

## allows(searchPermissions[])

Returns `true` if all `searchPermissions` are matched by the permission, otherwise returns `false`.

```js
import { permission } from 'url-permissions';

// Basic examples
permission('/articles:read').allows('/articles:read'); // true
permission('/articles:read,update').allows('/articles:read'); // true
permission('/articles:crud').allows('/articles:read,update'); // false
permission('/articles:read,update').allows('/articles:crud'); // false
permission('/articles:read').allows(['/articles:read', '/articles:update']); // false
permission('/articles/article-1:read').allows('/articles:read'); // false
permission('/articles:read').allows('/articles/article-1:read'); // false
permission('/articles:read,update').allows('/articles:read', '/articles:update'); // true
permission('/articles:read').allows('/articles:read', '/articles:update'); // false

// Query parameter examples
permission('/articles:read').allows('/articles?author=user-1:read'); // true
permission('/articles?author=user-1:read').allows('/articles:read'); // false
permission('/articles?author=user-1:read').allows('/articles?author=user-1&status=draft:read'); // true
permission('/articles?author=user-1&status=draft:read').allows('/articles?author=user-1:read'); // false

// Wildcards * and **
permission('/articles:read').allows('/art*cles:read'); // true
permission('/articles/article-1:read').allows('/articles/*:read'); // true
permission('/articles?author=user-2:read').allows('/articles/*:read'); // false
permission('/articles?author=user-2:read').allows('/articles/*:read'); // false
permission('/articles:read').allows('/articles/*:read'); // false
permission('/articles/*:read').allows('/articles/article-1/comments:read'); // false
permission('/articles/**:read').allows('/articles/article-1/comments:read'); // true

// Privilege aliases
permission('/articles:crud').allows('/articles:crud'); // true
permission('/articles:crud').allows('/articles:read'); // false
permission('/articles:read').allows('/articles:crud'); // false
```

## path([ path ])

Gets or sets the permission path.

```js
const perm = permission('/articles:read');

perm.path(); // "/articles"
perm.path('/users');
perm.path(); // "/users"
```

## parameters([ parameters ])

Gets or sets the permission parameters.

```js
const perm = permission('/articles?attr1=test:read');

perm.parameters(); // { attr1: 'test' }
perm.parameters({ attr1: 'test2', attr2: 'test3' });
perm.parameters(); // { attr1: 'test2', attr2: 'test3' }
```

## privileges([ privileges ])

Sets or returns priviliges

`privileges` can be either an array or comma-separated string with privilege names or their bitmasks. For example, `13`, `read`, `owner`, and `read,update,3` are valid.

When `privileges` are not defined method returns the bitmask of all privileges.

```js
const perm = permission('/articles:read');

perm.privileges(); // 1
perm.privileges('crud,own');
perm.privileges(); // 15 | 32 = 47
perm.privileges(['crud', 'manage', 'owner']);
perm.privileges(); // 15 | 16 | 63 = 63
```

## hasPrivilege(privilege)

Return `true` when permission has specified privilege(s).

`privileges` can be either an array or comma-separated string with privilege names or their bitmasks.

```js
const perm = permission('/articles:crud');

perm.hasPrivilege('read'); // true
perm.hasPrivilege(['read', 'create', 'update']); // true
perm.hasPrivilege('crud'); // true
perm.hasPrivilege('crud,read,create'); // true
perm.hasPrivilege('admin'); // false
perm.hasPrivilege('unknown'); // throws error
```

## hasPrivileges()

Alias of `hasPrivilege()`.

## grantPrivileges()

Gets array with privilege names that enable granting privileges. Returns empty array when none of permission's privileges are grant privileges.

```js
const perm = permission('/articles:read,manage,64');

perm.grantPrivileges(); // ['manage', 'admin']
```

## mayGrant(newPermission [, granteePermissions])

Returns `true` if permission allows granting `newPermission` to grantee.

In the grant process we distinguish two roles:

- *grantor*: the person granting or revoking a permission.
- *grantee*: the person receiving or losing a permission.

To grant a permission the grantor must have **manage** an/or **super** privilege.

- **manage** allows granting or revoking **crud** privileges from anyone without **manage** and **super** privilege.
- **super** allows granting and revoking permissions from anyone.

The grantor/grantee privileges can be customized using `permission.config()`.

```js
import { permission } from 'url-permissions';

permission('/articles:manage').mayGrant('/articles:read', []); // true
permission('/articles:manage').mayGrant('/articles:read', ['/articles:delete']); // true
permission('/articles:manage').mayGrant('/articles:read', ['/articles:admin']); // true
permission('/articles:manage').mayGrant('/articles:manage', ['/articles:manage']); // false
permission('/articles:manage').mayGrant('/articles:read', ['/unrelated:admin']); // true

permission('/articles:admin').mayGrant('/articles/article-1:read', ['/articles:manage']); // true
permission('/articles:admin').mayGrant('/articles/article-1:read', ['/articles:admin']); // true
```

## mayRevoke(permission [, granteePermissions])

Returns `true` if permission allows revoking `permission` to grantee.



```js
import { permission } from 'url-permissions';

permission('/articles:manage').mayRevoke('/articles:read', []); // true
permission('/articles:manage').mayRevoke('/articles:read', ['/articles:admin']); // false
permission('/articles:manage').mayRevoke('/articles:manage', ['/articles:manage']); // false

permission('/articles:admin').mayRevoke('/articles/article-1:read', ['/articles:manage']); // true
permission('/articles:admin').mayRevoke('/articles/article-1:read', ['/articles:admin']); // true
```

## toObject()

Returns an object representation of an URL permission containing three properties: `url`, `attributes` and `privileges`.

```js
import { permission } from 'url-permissions';

permission('/articles/*?author=user-1,user-2&flag=true:crud').toObject()
// {
//   path: '/articles/*',
//   attributes: {
//     author: ['user-1', 'user-2'],
//     flag: ['true'],
//   },
//   privileges: 15,
// }
```

## toString()

Returns a string representation of an URL Permission. Privileges are always represented by their bitmask.

```js
import { permission } from 'url-permissions';

permission('/articles/*?author=user-1:crud').toString()
// '/articles/*?author=user-1:15'
```

## clone()

Returns a clone of the permission.

```js
const a = permission('/articles:read');
const b = a.clone();
a.privileges(['u']);
b.privileges(); // ['r']
```

Alternatively you can do this:

```js
const a = permission('/articles:read');
const b = permission(a); // Clone a
```

## validate()

Static method to checks if url permission string is valid.

```js
import { permission } from 'url-permissions';

permission.validate('/articles?author=1,2:crud,manage'); // true
permission.validate('/articles?author=1,2'); // false, no privileges
permission.validate('/articles:unknown'); // false, unknown privilege
permission.validate('?author=user-1:create'); // false, no path
```

## config(options)

Static method to change global config options. Changing global config doesn't affect existing instances.

```js
import { permission } from 'url-permissions';

permission.config({
  privileges: {
    read: 1,
    create: 2,
    update: 4,
    delete: 8,
    crud: 15,
    manage: 16,
    manager: 31,
    own: 32,
    owner: 63,
    super: 64,
    admin: 127,
  },
  grantPrivileges: {
    manage: 15, // crud
    owner: 63, // owner + all the rest
    super: 127, // admin
  },
});
```

### Grant privileges

The config option `grantPrivileges` must be an object of key-value pairs where each key is a privilege name and each value a bitmask specifying which privileges it is allowed to grant.

Anyone with grant privileges is allowed to grant its privileges to grantees without any grant privileges. However, if a grantee does have grant privileges its grant privileges must be covered by those of the grantor.

Example:

```js
import { permission } from 'url-permissions';

permission.config({
  privileges: {
    a: 1,
    x: 2,
    y: 4,
    z: 8,
  },
  grantPrivileges: {
    x: 1,
    y: 3, // 1 | 2 => a and x are allowed
    z: 9, // 1 | 8 => a and z are allowed
  },
});

permission('/articles:x').mayGrant('/articles:a'); // true
permission('/articles:x').mayGrant('/articles:a', ['/articles:x']); // false
permission('/articles:y').mayGrant('/articles:a', ['/articles:x']); // true
permission('/articles:y').mayGrant('/articles:x', ['/articles:x']); // true
permission('/articles:y').mayGrant('/articles:a', ['/articles:y']); // false
permission('/articles:z').mayGrant('/articles:a', ['/articles:z']); // true
```

Notice the last example where `z` may grant a permission to a grantee with `z`, whereas an `y` may not grant the same permission to another `y`. We'll leave it to you to figure out why.

# permissions(perms)

The `permissions(perms)` function enables verifying a collection of permissions.

Valid `perms` are permission strings, `permission()` objects, or arrays of either of those.

## allows(searchPermissions[])

Returns `true` if all `searchPermissions` are matched by any single or a combination of `permissions`.

This method intelligently handles combination and products of parameter values and privileges.

```js
import { permissions } from 'url-permissions';

permissions('/articles:read', '/articles:update').allows('/articles:ru'); // true
permissions('/articles/*:read', '/articles/*:update').allows('/articles/article-1:ru'); // true
permissions('/articles?author=user1:read', '/articles?author=user2:read')
  .allows('/articles?author=user1,user2:read'); // true
permissions('/articles?author=user1:read', '/articles?author=user2:update')
  .allows('/articles?author=user1,user2:read,update'); // false
```

When using comma-separated parameter values each value should be considered a separate permission that must be allowed:

```js
permissions('/articles?author=user-1:read', '/articles?author=user-2:read')
  .allows('/articles?author=user-1,user-2&status=published:read'); // true

// ... is equivalent to:
permissions('/articles?author=user-1:read', '/articles?author=user-2:read')
  .allows([
    '/articles?author=user-1&status=published:read',
    '/articles?author=user-2&status=published:read',
  ]);
```

## mayGrant(newPermission [, granteePermissions])

Returns `true` if any or a combination of this collection's permissions allow granting `newPermission` to grantee.

For a better understanding of how granting work, see `permission().mayGrant(...)`.

```js
import { permissions } from 'url-permissions';

permission('/articles:read', '/articles:m').mayGrant('/articles:read'); // true

permissions('/articles?author=user-1:owner', '/articles?author=user-2:owner')
  .mayGrant('/articles?author=user-1,user-2:read', ['/articles:read']); // true

permissions('/articles?author=user-1:manage', '/articles?author=user-2:manage')
  .mayGrant('/articles?author=user-1,user-2:read', ['/articles:owner']); // false
```

## mayRevoke(removePermission [, granteePermissions])

An alias of `permissions().mayGrant()`.

Returns `true` if any or a combination of this collection's permissions allow revoking `removePermission` from grantee.

## permissions([ permissions ])

Gets or sets permissions.

`permissions` can be either a string or an array of permissions.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

# [2.0.0] - 2016-01-15
### Added
- Use bitmasks to specify which privileges are allowed
- Added `administrator` privilege with bitmask for all other privileges
- Added `own` privilege enabling `owner`-unique operations

### Changed
- Changed aliases to normal privileges
- Renamed `all` alias to `crud` and made it a normal privilege
- Removed privilege (one-character) identifiers
- Renamed `super` privilege to `admin`

## [1.1.0] - 2016-01-07
### Added
- `permission().hasPrivilege()` method
- `permission().hasPrivileges()` alias

## [1.0.0] - 2016-01-07
### Fixed
- Fix parameters being validated in the wrong order
- `parameters()` now processes `{ attr: 'value1,value2' }` into `{ attr: ['value1', 'value2' ]}` rather than `{ attr: 'value1,value2' }`

### Changed
- Refactored test files setup
- Refactor `export default permission` to `export { permission }`
- Changed "slow" test indicator from default to 10ms
- `permission()` constructor now accepts an URLPermission object and clones it
- `mayRevoke()` is now yet again an alias of `mayGrant()`.
- `mayGrant()` no longer allows users with higher privileges to grant permissions.

### Added
- `permissions()` function and companion class `URLPermissions` evaluate a collection of permissions
- `permission().clone()` method

## [0.4.0] - 2016-12-30
### Fixed
- Fixed basic permissions being incorrectly allowed, e.g. `/articles:read,update` would "allow" `/articles:all`.

### Changed
- `allows()` now accepts permissions as varargs.
- `allows()` now evaluates permissions as AND rather than OR.

### Added
- Add `toString()` method.

## [0.3.0] - 2016-12-30
### Changed
- Renamed `matches()` to `allows()`

## [0.2.0] - 2016-12-27
### Changed
- `mayGrant()` now allows granting basic privileges (crud) to anyone.
- `mayRevoke()` is no longer an alias of `mayGrant()` and does not allow revoking basic privileges (crud) from anyone with a “manage” or “super” privilege

## [0.1.1] - 2016-12-26
### Changed
- URL Permission constructor throws error when input is not a string.

### Fixed
- Fixed bug where missing parameters yield `false` rather than `true` for matchers

## [0.1.0] - 2016-12-22
### Added
- `validate()` method

## [0.0.2] - 2016-12-17
Complete rewrite.

### Added
- `config()` method
- `matches()` method
- `mayGrant()` method

## [0.0.1] - 2016-11-15
### Added
- `verify()` method
- `parse()` method
- `config()` method


test/config.js
import { expect } from 'chai';
import { permission } from '../src/index';

beforeEach(() => {
  // Reset global config
  permission.config(false);
});

describe('config(...)', function() {
  this.slow(10);

  describe('grant privileges', () => {
    it('should throw an error when grant privileges is not an object', () => {
      const func = () => permission.config({
        grantPrivileges: { a: { c: 't', r: 'r' } },
      });
      expect(func).to.throw('Grant privilege \'a\' must specify a valid number');
    });

    it('should throw an error when grant privilege identifier does not exist', () => {
      const func = () => permission.config({
        grantPrivileges: { all: ['c', 'r', 'u', 'd'] },
      });
      expect(func).to.throw('Grant privilege \'all\' must specify a valid number');
    });

    it('should configure all grant privileges', () => {
      permission.config({
        grantPrivileges: { create: 5 },
      });
      expect(permission.config().grantPrivileges).to.deep.equal({ create: 5 });
    });
  });

  describe('privileges', () => {
    it('should throw an error when privileges is not an object', () => {
      const func = () => permission.config({ privileges: ['a'] });
      expect(func).to.throw('Privileges must an object');
    });

    it('should throw an error when privilege does not specify a number', () => {
      const func = () => permission.config({ privileges: { abc: 'ef' } });
      expect(func).to.throw('Privilege identifier \'abc\' must be a number');
    });

    it('should configure privileges globally', () => {
      permission.config({
        privileges: {
          read: 1,
          create: 2,
        },
      });
      expect(permission.config().privileges).to.deep.equal({
        read: 1,
        create: 2,
      });
      expect(permission('/articles:2').allows('/articles:create')).to.equal(true);
      const func = () => permission('/articles:delete').allows('/articles:delete');
      expect(func).to.throw(/Privilege 'delete' does not exist/);
    });
  });

  it('should not affect existing instances', () => {
    const perm = permission('/articles:crud,admin');
    permission.config({
      privileges: {
        read: 1,
        create: 2,
      },
      grantPrivileges: {
        create: 3,
      },
    });
    expect(perm.grantPrivileges()).to.deep.equal(['admin']);
    expect(perm.grantsAllowed()).to.equal(127);
    expect(perm.privileges()).to.deep.equal(15 + 64);
    perm.privileges(16);
    expect(perm.privileges()).to.deep.equal(16);
  });
});


test/validate.js
import { expect } from 'chai';
import { permission } from '../src/index';

describe('validate(...)', function() {
  this.slow(10);

  it('should return true when string is valid', () => {
    expect(permission.validate('/articles?author=1,2:crud,manage')).to.equal(true);
    expect(permission.validate('https://examp.le/articles?author=1,2:crud,manage')).to.equal(true);
    expect(permission.validate('/articles:crud,manage,admin')).to.equal(true);
  });

  it('should return false string with global config', () => {
    expect(permission.validate(false)).to.equal(false);
    expect(permission.validate('/articles?author=1,2')).to.equal(false);
    expect(permission.validate('/articles:unknown')).to.equal(false);
    expect(permission.validate('/articles:crudmanage')).to.equal(false);
    expect(permission.validate(':create')).to.equal(false);
    expect(permission.validate('?author=user-1:create')).to.equal(false);
  });
});


src/validate.js
import _ from 'lodash';
import config from './config';
import { isNumeric } from './util';

/**
 * Validates privilege string.
 *
 * A privilege string can contain any combination of privilege names and numbers.
 *
 * Example:
 * ```
 * validatePrivileges('crud,manage'); // true
 * validatePrivileges('5,read'); // true
 * validatePrivileges('invalid'); // false
 * ```
 */
function validatePrivileges(privilegeString) {
  const settings = config();
  const privileges = privilegeString.split(',');
  const privilegeNames = Object.keys(settings.privileges);
  const privilegeNumbers = _.values(settings.privileges);

  for (const privilege of privileges) {
    if (isNumeric(privilege)) {
      if (!privilegeNumbers.includes(privilege)) return false;
    } else if (!privilegeNames.includes(privilege)) {
      return false;
    }
  }
  return true;
}

/**
 * Validates an URL permission string.
 */
export default (permission) => {
  if (!_.isString(permission)) {
    return false;
  }

  // Check if string contains privilege delimiter
  const privilegeDelimiterIndex = permission.lastIndexOf(':');
  if (privilegeDelimiterIndex === -1) {
    return false;
  }
  const url = permission.substring(0, privilegeDelimiterIndex);

  // Check if path length >= 1
  const attributeDelimiterIndex = url.indexOf('?');
  if (attributeDelimiterIndex === 0 || url.length === 0) {
    return false;
  }

  // Check if all privileges are valid aliases, action names or identifiers
  const validPrivilege = validatePrivileges(permission.substring(privilegeDelimiterIndex + 1));
  if (!validPrivilege) return false;

  return true;
};


gulpfile.babel.js
/* eslint import/no-extraneous-dependencies: 'off' */
/* eslint arrow-body-style: 'off' */

import gulp from 'gulp';
import eslint from 'gulp-eslint';
import babel from 'gulp-babel';
import sourcemaps from 'gulp-sourcemaps';
import del from 'del';

const paths = {
  src: 'src',
  dest: 'lib',
  lint: ['gulpfile.babel.js', 'src/**/*.js', 'test/**/*.js'],
};

gulp.task('clean', () => del(`${paths.dest}/*`));

gulp.task('build', ['clean'], () => {
  return gulp.src(`${paths.src}/**/*.js`)
    .pipe(sourcemaps.init())
    .pipe(babel())
    .pipe(sourcemaps.write('.'))
    .pipe(gulp.dest(paths.dest));
});

gulp.task('lint', () => {
  return gulp.src(paths.lint)
    .pipe(eslint())
    .pipe(eslint.format());
});

gulp.task('watch', ['build', 'lint'], () => {
  gulp.watch(paths.lint, ['build', 'lint']);
});

gulp.task('default', ['build', 'lint']);


src/globToRegex.js
/**
 * Parses * and ** in glob pattern.
 */
export default (pattern) => {
  let result = '';
  let char = '';

  for (let i = 0, len = pattern.length; i < len; i += 1) {
    char = pattern[i];

    switch (char) {
      case '$':
      case '.':
      case '=':
      case '!':
      case '^':
      case '+':
      case '|':
      case '[':
      case ']':
      case '(':
      case ')':
      case '?':
      case '{':
      case '}':
      case '/':
      case '\\':
        result += `\\${char}`; // Escape reserved regex character
        break;
      case '*': {
        // A globstar is `abc/**/xyz`, `abc/**`, `**/abc` but not `/a**/`.
        // To determine if it is a ** rather than *, consume all consecutive "*"
        // and find the chars directly before and after '**'
        const prev = pattern[i - 1];
        let multipleStars = false;
        while (pattern[i + 1] === '*') {
          multipleStars = true;
          i += 1;
        }
        const next = pattern[i + 1];
        if (multipleStars && (!prev || prev === '/') && (!next || next === '/')) {
          result += '([^/]*(/|$))*';
          i += 1; // Move past "/"
        } else {
          result += '[^/]*';
        }
        break;
      }
      default:
        result += char;
    }
  }

  return new RegExp(`^${result}$`);
};


src/URLPermissions.js
import _ from 'lodash';
import URLPermission from './URLPermission';

/**
 * Models multiple permissions.
 */
export default class URLPermissions {
  /**
   * Creates new URLPermission instance based on a permission string.
   */
  constructor(...perms) {
    this._permissions = _.flatten(perms).map(e => new URLPermission(e));
  }

  /**
   * Sets or retrieves permissions.
   */
  permissions(permissions) {
    if (permissions !== undefined) {
      this._permissions = _.flatten(perms).map(e => new URLPermission(e));
      return this;
    }
    return this._permissions;
  }

  /**
   * Returns `true` when all specified permissions are covered by at least one
   * of the permissions.
   *
   * `permissions` can be either permission strings or an URLPermission object.
   *
   * Throws an error when any permission is invalid.
   */
  allows(...perms) {
    // Transform each permission string to an object.
    const permissions = _.flatten(_.flatten(perms).map(e => {
      return this._unwindParameters(e);
    }));

    // For each permission remove any privilege that is covered by any of our
    // own permissions.
    const ourPermissions = this.permissions();
    return _.every(permissions, (permission) => {
      let i = 0;
      while (permission.privileges() > 0 && i < ourPermissions.length) {
        if (ourPermissions[i].matchUrl(permission)) {
          permission.privileges(permission.privileges() & ~ourPermissions[i].privileges()); // Remove flag
        }
        i++;
      }
      return permission.privileges() === 0;
    });
  }

  /**
   * Returns `true` if any or a combination of this collection's permissions
   * allow granting `newPermission` to grantee.
   */
   mayGrant(newPermission, granteePermissions = []) {
     return this._mayGrantOrRevoke('mayGrant', newPermission, granteePermissions);
   }

  /**
   * Returns `true` if any or a combination of this collection's permissions
   * allow revoking `permission` from grantee.
   */
   mayRevoke(permission, granteePermissions = []) {
     return this._mayGrantOrRevoke('mayRevoke', permission, granteePermissions);
   }

   /**
    * This private function contains the common logic of `mayGrant` and
    * `mayRevoke`. Specify `functionName` either `mayGrant` or `mayRevoke`.
    */
   _mayGrantOrRevoke(functionName, newPermission, granteePermissions = []) {
     const newPerms = _.flatten(this._unwindParameters(newPermission));
     const ourPerms = this.permissions();
     return _.every(newPerms, (newPerm) => {
       return _.some(ourPerms, (ourPerm) => ourPerm[functionName](newPerm, granteePermissions));
     });
   }

  /**
   * Returns an array with permissions for each and every parameter value.
   *
   * For example:
   * ```
   * _unwindPermission('/articles?var1=foo1,foo2&var2=foo3,foo4:read,update');
   * // [
   * //   "/articles?var1=foo1&var2=foo3:read,update",
   * //   "/articles?var1=foo1&var2=foo4:read,update",
   * //   "/articles?var1=foo2&var2=foo3:read,update",
   * //   "/articles?var1=foo2&var2=foo4:read,update",
   * // ]
   * ```
   */
  _unwindParameters(permission) {
    const perm = new URLPermission(permission);
    const params = perm.parameters();
    if (!params) return [perm]; // Simply return, nothing to unwind

    // Recursive function that returns an array with one object for each unique
    // path in the `params` object.
    const unwind = (keys, keyIndex) => {
      const result = [];
      params[keys[keyIndex]].forEach((value) => {
        const newValue = { [keys[keyIndex]]: value };
        if (keyIndex === keys.length - 1) {
          // We've reached the end of the tree, simply return the new value
          result.push(newValue);
        } else {
          // We've not yet reached the end of the tree, append each child to new value
          unwind(keys, keyIndex + 1).forEach((child) => {
            result.push(Object.assign({}, newValue, child));
          });
        }
      });
      return result;
    }

    // For each path, generate a new permission with unique parameters
    return unwind(Object.keys(params), 0).map((unwindedParams) => {
      return perm.clone().parameters(unwindedParams);
    });
  }
}


src/config.js
import _ from 'lodash';
import { isNumeric } from './util';

// Contains default global config
const defaultConfig = {
  privileges: {
    read: 1,
    create: 2,
    update: 4,
    delete: 8,
    crud: 15,
    manage: 16,
    manager: 31,
    own: 32,
    owner: 63,
    admin: 64,
    administrator: 127,
  },
  grantPrivileges: {
    manage: 15, // crud
    own: 63, // owner + all the rest
    admin: 127, // admin
  },
};

const config = _.cloneDeep(defaultConfig);

/**
 * Sets or retrieves privileges which allow granting permissions.
 */
function grantPrivileges(newGrantPrivileges = null) {
  if (newGrantPrivileges === false) {
    config.grantPrivileges = _.cloneDeep(defaultConfig.grantPrivileges);
  } else if (newGrantPrivileges) {
    Object.keys(newGrantPrivileges).forEach((key) => {
      if (!isNumeric(newGrantPrivileges[key])) {
        throw new Error(`Grant privilege '${key}' must specify a valid number`);
      }
    });

    // If config is valid assign override existing config
    config.grantPrivileges = _.cloneDeep(newGrantPrivileges);
  }
  return config.grantPrivileges;
}

/**
 * Sets or retrieves global privilege config.
 */
function privileges(newPrivileges = null) {
  if (newPrivileges === false) {
    config.privileges = _.cloneDeep(defaultConfig.privileges);
  } else if (newPrivileges) {
    if (!_.isPlainObject(newPrivileges)) {
      throw new Error('Privileges must an object');
    }
    Object.keys(newPrivileges).forEach((key) => {
      if (!isNumeric(newPrivileges[key])) {
        throw new Error(`Privilege identifier '${key}' must be a number`);
      }
    });
    config.privileges = _.cloneDeep(newPrivileges);
  }
  return config.privileges;
}

/**
 * Retrieves or sets configuration.
 *
 * When `false` resets config to default.
 */
export default (options) => {
  if (options === false) {
    privileges(false);
    grantPrivileges(false);
  } else if (options) {
    if (options.privileges) privileges(options.privileges);
    if (options.grantPrivileges) grantPrivileges(options.grantPrivileges);
  }
  return config;
};
